src/staticdata.c

```diff
@@ -1799,6 +1799,44 @@ static void jl_set_nroots_sysimg(void)
 static void jl_init_serializer2(int);
 static void jl_cleanup_serializer2(void);
 
+// NOTE: currently protected by the codegen lock, since it is only used there
+jl_array_t *jl_global_roots_table;
+
+JL_DLLEXPORT int jl_is_globally_rooted(jl_value_t *val)
+{
+    if (jl_is_concrete_type(val) || jl_is_bool(val) || jl_is_symbol(val) ||
+            val == (jl_value_t*)jl_any_type || val == (jl_value_t*)jl_bottom_type || val == (jl_value_t*)jl_core_module)
+        return 1;
+    if (val == ((jl_datatype_t*)jl_typeof(val))->instance)
+        return 1;
+    return 0;
+}
+
+JL_DLLEXPORT jl_value_t *jl_as_global_root(jl_value_t *val)
+{
+    if (jl_is_globally_rooted(val))
+        return val;
+    if (jl_is_uint8(val))
+        return jl_box_uint8(jl_unbox_uint8(val));
+    if (jl_is_int32(val)) {
+        int32_t n = jl_unbox_int32(val);
+        if ((uint32_t)(n+512) < 1024)
+            return jl_box_int32(n);
+    }
+    else if (jl_is_int64(val)) {
+        uint64_t n = jl_unbox_uint64(val);
+        if ((uint64_t)(n+512) < 1024)
+            return jl_box_int64(n);
+    }
+    jl_value_t *rval = jl_eqtable_getkey(jl_global_roots_table, val, NULL);
+    if (rval)
+        return rval;
+    JL_GC_PUSH1(&val);
+    jl_global_roots_table = jl_eqtable_put(jl_global_roots_table, val, jl_nothing, NULL);
+    JL_GC_POP();
```

Comment:
you have a function marked as a safepoint earlier in the function (jl_is_globally_rooted–though it is not), so analysis shouldn't think this is needed
```suggestion
    jl_global_roots_table = jl_eqtable_put(jl_global_roots_table, val, jl_nothing, NULL);
```